pkg/connector/mapping.go

```diff
@@ -0,0 +1,348 @@
+	"maunium.net/go/mautrix/bridgev2/database"
+	"maunium.net/go/mautrix/bridgev2/networkid"
+	bridgeEvt "maunium.net/go/mautrix/event"
+
+	"go.mau.fi/mautrix-twitter/pkg/twittermeow/data/types"
+	"go.mau.fi/mautrix-twitter/pkg/twittermeow/methods"
+)
+
+func (tc *TwitterClient) MessagesToBackfillMessages(ctx context.Context, messages []types.Message, conv types.Conversation) ([]*bridgev2.BackfillMessage, error) {
+	backfilledMessages := make([]*bridgev2.BackfillMessage, 0)
+	selfUserId := tc.client.GetCurrentUserID()
+	for _, msg := range messages {
+		backfilledMessage, err := tc.MessageToBackfillMessage(ctx, msg, conv, selfUserId)
+		if err != nil {
+			return nil, err
+		}
+		backfilledMessages = append(backfilledMessages, backfilledMessage)
+	}
+
+	return backfilledMessages, nil
+}
+
+func (tc *TwitterClient) MessageToBackfillMessage(ctx context.Context, message types.Message, conv types.Conversation, selfUserId string) (*bridgev2.BackfillMessage, error) {
+	messageReactions, err := tc.MessageReactionsToBackfillReactions(message.MessageReactions, selfUserId)
+	if err != nil {
+		return nil, err
+	}
+
+	sentAt, err := methods.UnixStringMilliToTime(message.MessageData.Time)
+	if err != nil {
+		return nil, err
+	}
+
+	partId := networkid.PartID("")
+	parts := make([]*bridgev2.ConvertedMessagePart, 0)
+
+	textPart := &bridgev2.ConvertedMessagePart{
+		ID:   partId,
+		Type: bridgeEvt.EventMessage,
+		Content: &bridgeEvt.MessageEventContent{
+			MsgType: bridgeEvt.MsgText,
+			Body:    message.MessageData.Text,
+		},
+	}
+
+	replyData := message.MessageData.ReplyData
+	var replyTo *networkid.MessageOptionalPartID
+	if replyData.ID != "" {
+		replyTo = &networkid.MessageOptionalPartID{
+			MessageID: networkid.MessageID(replyData.ID),
+			PartID:    &partId,
+		}
+	}
+
+	if message.MessageData.Attachment != nil {
+		portal, err := tc.connector.br.GetPortalByKey(ctx, tc.MakePortalKey(conv))
+		if err != nil {
+			return nil, err
+		}
+
+		convertedAttachmentPart, indices, err := tc.TwitterAttachmentToMatrix(ctx, portal, tc.userLogin.Bridge.Matrix.BotIntent(), message.MessageData.Attachment)
+		if err != nil {
+			return nil, err
+		}
+
+		RemoveEntityLinkFromText(textPart, indices)
+		parts = append(parts, convertedAttachmentPart)
+	}
+
+	if len(textPart.Content.Body) > 0 {
+		parts = append(parts, textPart)
+	}
+
+	return &bridgev2.BackfillMessage{
+		ConvertedMessage: &bridgev2.ConvertedMessage{
+			ReplyTo: replyTo,
+			Parts:   parts,
+		},
+		Sender: bridgev2.EventSender{
+			IsFromMe: message.MessageData.SenderID == selfUserId,
+			Sender:   networkid.UserID(message.MessageData.SenderID),
+		},
+		ID:        networkid.MessageID(message.MessageData.ID),
+		Timestamp: sentAt,
+		Reactions: messageReactions,
+	}, nil
+}
+
+func RemoveEntityLinkFromText(msgPart *bridgev2.ConvertedMessagePart, indices []int) {
+	start, end := indices[0], indices[1]
+	msgPart.Content.Body = msgPart.Content.Body[:start-1] + msgPart.Content.Body[end:]
```

Comment:
Are the indexes really utf-8 bytes? I thought most networks use utf-16 indexes

Reply:
![CleanShot 2024-10-17 at 08 09 55](https://github.com/user-attachments/assets/0baa8001-5258-4fda-bd2a-b43c6f79afc6)

looks utf8 to me 🤔